### ISSUE TITLE
Improve website performance as audited by LightHouse

### ISSUE THREAD

**Issue by TimvdLippe** [WISVCH/website#18]
I ran a LightHouse audit on the new website and the results are a definitive improvement over the old website! However, there are still some areas for improvement. 

See the following JSON file for the full audit [ch.tudelft.nl-20171125T151003.log](https://github.com/WISVCH/website/files/1503129/ch.tudelft.nl-20171125T151003.log) (it is named a `.log`, but it is actually `.json`. You will need to rename the file back. See https://github.com/isaacs/github/issues/1130 for more details). You can view a more friendly HTML format by uploading this file to https://googlechrome.github.io/lighthouse/viewer/

* Most notably, images are a big factor, which is already tracked in #16, as 2.1 out of 2.3MB is just images. It seems that user-uploaded content does not properly go through a minification pipeline. A very quick Google resulted in https://nl.wordpress.org/plugins/ewww-image-optimizer/ which might be a solution? (I have no experience with Wordpress, so this is probably better your call)
* Images are not lazy-loaded. Impact is less big than the previous point, but lazy-loading might still be a good thing to do.
* Images that are uploaded are also too big for their actual usage. This means that potentially a 4K image is uploaded which is rendered in a 100x200 box. I know that imagemagick has an option to resize, so maybe that Wordpress plugin has an option as well?
* The website seems heavily depended on jQuery as well as heavy bundling of all assets. This means that the full load is blocked on all JS and CSS before it can render any content. See the Critical Request Chain in the report for more information. Ideally this should be as close to 1 as possible.
  * It seems that the jQuery is loaded by `wp_head()`? Is it an integral part of the theme or can it be lazy-loaded?
  * The version of jQuery (1.12.4) is also more than 1,5 year old (https://en.wikipedia.org/wiki/JQuery#Release_history) while 3.0 has been out for while. However, you are also loading https://github.com/jquery/jquery-migrate which only seems necessary if you are actually using jQuery 3.x. I think that either jquery-migrate can be removed, or that jQuery should be updated to 3.0 (if it can't be removed)
  * There are 3 `woff2` fonts that are being loaded. Is this something we can control? It seems that only 1 `woff` and 1 `woff2` are required
* There are some small Accessibility issues, seems rather straightforward to add.
* A lot of the buttons and text violate the color contrast check (which seems the case throughout the theme). While this issue is not the most important one for now, it might be nice to update the theme to make it accessible to everyone.
* When loading on mobile, it seems that the background color of all sessions is rendered first, only then the images pop in. I think this is related to the issue of not specifying the image size, so probably you can fix this issue when you are working on the performance stuff anyways.

All in all I think the new theme is a huge improvement, especially on mobile. With these last points, the website is blazingly fast on probably every network. Congrats!

Feel free to poke me if you have any more questions. I am also happy to contribute PRs with fixes, if I can figure out how this Wordpress thingy works :smile:

**Comment by svenvanhal** [comment]
Hi Tim, thanks for performing this analysis. I see LightHouse is a little biased towards Google technologies, but helpful anyway.

- The image problem is noted and being fixed. WordPress does in fact generate smaller images, but the theme loads the wrong size in some places. See WISVCH/rechallenge#8. We could look at minifying the resized images, but we'll have to see if that's worth the hassle. Theme assets have already been optimized. 
- Lazy loading seems like a nice feature to implement in the future.
- Small images in big boxes: related to the first problem. Most images are slightly bigger than their box on purpose (for high DPI screens). 
- WordPress is bundled with jQuery 1.12.4 and Migrate for compatibility reasons, but we could use 3.x for the front-end anyways. I also think the theme-specific javascript is slightly gebeund at the moment, so when upgrading to Foundation 6.4 (#8), we should look at the javascript in its entirety.
- About the fonts: we use three font variants (heading, body and body bold) and a fourth icon font (which only includes the necessary glyphs), so that makes four in total. The icon font is more versatile than a sprite (and probably smaller) and the total font pageload (vs "visual gains") isn't that big, so I'm fine with the current situation.
- The main accessibility issues are about color contrast, but it looks like they're off by just a little. In practice, I think the impact is rather small, so I'll take a look at it after fixing the more important stuff.
- Do you mean the background of the slider or the navigation menu? The latter is quite annoying (will take a look at that), but I don't really see how we can fix the slider. Maybe fade the first images in when it has loaded?

Thanks!

**Comment by TimvdLippe** [comment]
I just found out that minification is not 100% fixed as well.

- https://ch.tudelft.nl/wp-content/themes/rechallenge/assets/js/scripts.js still contains JS comments
- https://ch.tudelft.nl/ has no minified HTML
- https://ch.tudelft.nl/wp-content/themes/rechallenge/assets/fonts/font-wisvch.css is not minified as well

For CSS and JS you can use default frontend minifiers like https://github.com/css/csso and https://github.com/babel/minify. For the HTML one generated by PHP, it seems that https://packagist.org/packages/zaininnari/html-minifier might be a possibility?

**Comment by svenvanhal** [comment]
scripts.js is already being minified (my mistake): https://github.com/WISVCH/rechallenge/blob/66c4b31ccc6119ed99e9192bc09f4666b10e1f18/lib/Assets.php#L36

Good catch @ the font css, I think we can inline the file in the other stylesheet, since it's now @import'ed after app.css is loaded.

The HTML is indeed not minified, but also not static so we would need some kind of WordPress plugin for this. We could however look at caching and minifying the HTML in the process.

**Comment by TimvdLippe** [comment]
First result on Google is
https://nl.wordpress.org/plugins/minify-html-markup/ which seems actively
updated. Seems like an easy win :P

Ah nice about JS and sounds good for the CSS!

Op di 28 nov. 2017 om 19:20 schreef Sven van Hal <notifications@github.com>:

**Comment by svenvanhal** [comment]
True, but I don't know if that would actually make the site faster. Knowing WordPress, I think that the added overhead / processing time exceeds the gains.

**Comment by praseodym** [comment]
Caching sounds like a good idea though!
